connectors/utils.py

```diff
@@ -399,3 +401,60 @@ async def func_to_execute(*args, **kwargs):
         return func_to_execute
 
     return wrapper
+
+
+def ssl_context(certificate):
+    """Convert string to pem format and create a SSL context
+
+    Args:
+        certificate (str): certificate in string format
+
+    Returns:
+        ssl_context: SSL context with certificate
+    """
+    certificate = certificate.replace(" ", "\n")
+    certificate = " ".join(certificate.split("\n", 1))
+    certificate = " ".join(certificate.rsplit("\n", 1))
+    ctx = ssl.create_default_context()
+    ctx.load_verify_locations(cadata=certificate)
+    return ctx
+
+
+def encode(original_string):
+    """Performs encoding on the objects
+    containing special characters in their url, and
+    replaces single quote with two single quote since quote
+    is treated as an escape character
+
+    Args:
+        original_string(string): String containing special characters
+
+    Returns:
+        encoded_string(string): Parsed string without single quotes
+    """
+    encoded_string = urllib.parse.quote(original_string, safe="'")
+    return encoded_string.replace("'", "''")
+
+
+def get_expires_at(expires_in):
+    """Adds seconds to the current utc time.
+
+    Args:
+        expires_in (int): Seconds after which the token will expire.
+    """
+    expires_at = datetime.utcnow() + timedelta(seconds=expires_in)
+    # account for clock skew
+    expires_at -= timedelta(seconds=20)
```

Comment:
How did you come up with 20 seconds? Is this something that SharePoint recommends? Have you ever encountered a clock skew issue while working with Sharepoint?

Reply:
> How did you come up with 20 seconds?

Assuming that the servers are in sync with NIST time(https://www.nist.gov/pml/time-and-frequency-division/time-distribution/internet-time-service-its), then the difference should be negligible at around a couple of seconds only. keeping this in mind I chosed 20s

> Is this something that SharePoint recommends?Have you ever encountered a clock skew issue while working with Sharepoint?
   
No it's not receommend by sharepoint but we did this just for precautionary purpose. Suppose we have two auth servers.How do you know that they will have the same time? What if they are in fact off by a few minutes. One server will return that the token has expired and the other won't.
we don't need it it to be required here but if you have two auth servers then we can face clock skew issue.